### PR TITLE
[clang][FMV][AArch64] Improve streaming mode compatibility

### PR DESCRIPTION
**Pull request by labrinea**
* Allow arm-streaming if all the functions versions adhere to it.
* Allow arm-streaming-compatible if all the functions versions adhere to it.

When the caller needs to toggle the streaming mode all the function
versions of the callee must adhere to the same mode, otherwise the call
will yield a runtime error.

Imagine the versions of the callee live in separate TUs. The version
that is visible to the caller will determine the calling convention used
when generating code for the callsite. Therefore we cannot support
mixing streaming with non-streaming function versions. Imagine TU1 has a
streaming caller and calls foo._sme which is streaming-compatible. The
codegen for the callsite will not switch off the streaming mode. Then in
TU2 we have a version which is non-streaming and could potentially be
called in streaming mode. Similarly if the caller is non-streaming and
the called version is streaming-compatible the codegen for the callsite
will not switch on the streaming mode, but other versions may be
streaming.

### CODE REVIEW

**Comment by labrinea** [comment]
backport of [f8ae128](https://github.com/llvm/llvm-project/commit/f8ae128755777424cf4133e4e8e819c0bc08d2b1) on [llvm:release/19.x](https://github.com/llvm/llvm-project/tree/release/19.x).

**Comment by tru** [comment]
@AaronBallman @cor3ntin @erichkeane Should this be merged in the backport?

**Comment by tru** [comment]
Oh maybe we should have some arm64 person chime in as well.

**Comment by AaronBallman** [comment]
I'm not certain why this is necessary to backport; is this fixing a regression or completing some work started in Clang 19?

**Comment by labrinea** [comment]
> I'm not certain why this is necessary to backport; is this fixing a regression or completing some work started in Clang 19?

As explained on the description it is fixing a potential runtime error when mixing streaming with non-streaming function versions.

**Comment by AaronBallman** [comment]
> > I'm not certain why this is necessary to backport; is this fixing a regression or completing some work started in Clang 19?
> 
> As explained on the description it is fixing a potential runtime error when mixing streaming with non-streaming function versions.

We used to diagnose streaming functions that attempted to be multiversioned, this removes that restriction; that doesn't seem like a bug fix so much as improving a feature. I'm still not understanding why this is necessary to backport -- can you post a Compiler Explorer link that demonstrates the buggy behavior, maybe that would help me to understand a bit better?

**Comment by labrinea** [comment]
> We used to diagnose streaming functions that attempted to be multiversioned

but we did not diagnose a streaming caller of a versioned function, and since streaming compatible versions were not diagnosed either it was possible to mix calling conventions between versions:

- Streaming function calls streaming compatible version: https://godbolt.org/z/E6vnWd33x. Note that if bar._sve() was declared non streaming, then the call would be surrounded by smtop/smstart: https://godbolt.org/z/Wdf8oWoxh.
- In a separate translation unit, another version is defined (the default) with a different calling convention (non streaming). With llvm18 this compiles just fine: https://godbolt.org/z/GxzEGr6Yx
- The same as before in llvm trunk yields a semantic error: https://godbolt.org/z/fzbb5hTGb

**Comment by AaronBallman** [comment]
Ah thank you for the explanation -- what was tripping me up was the changes in `clang/test/Sema/aarch64-sme-func-attrs.c` where we used to issue diagnostics. Okay, this makes more sense as to why we'd want to backport it then, so I retract my objection.

**Review by AaronBallman — APPROVED** [review]
The changes LGTM -- the precommit CI failures seem to be unrelated, but it would be good to get a second opinion.

After the pick lands, please be sure to add a release note so users know about the fix.

**Review by sdesmalen-arm — APPROVED** [review]
It would be good if this could make it into the LLVM 19 release.

**Comment by labrinea** [comment]
> @labrinea (or anyone else). If you would like to add a note about this fix in the release notes (completely optional). Please reply to this comment with a one or two sentence description of the fix. When you are done, please add the release:note label to this PR.

Streaming mode compatibilty has been improved for Multiversioned Functions which reside in separate translation units. All the versions of a function must have the same calling convention. If not, the compiler emits diagnostics.

**Pull request merged**
